### PR TITLE
Spark 2.x support

### PR DESCRIPTION
**Pull request by elyast**
Prepared patch to allow compilation against Spark 2.x, it is based on my contribution to DataVec: https://github.com/deeplearning4j/DataVec/pull/131

### CODE REVIEW

**Comment by elyast** [comment]
any progress on this ?

File: deeplearning4j-scaleout/dl4j-streaming/src/main/spark-2/org/deeplearning4j/streaming/pipeline/spark/StreamingContextUtils.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ */
+public class StreamingContextUtils {
+
+    public static void awaitTermination(JavaStreamingContext jssc, long timeout) {
+        try {
+            if(timeout < 0)
+                jssc.awaitTermination();
+            else
+                jssc.awaitTerminationOrTimeout(timeout);
+        } catch (InterruptedException e) {
```
Comment (huitseeker) [inline deeplearning4j-scaleout/dl4j-streaming/src/main/spark-2/org/deeplearning4j/streaming/pipeline/spark/StreamingContextUtils.java]: You can avoid duplicating this class and catch it as a general exception:
```java
try {
  jssc.awaitTermination();
} catch (Exception e) {
  if (e instanceof InterruptedException) {
    // handle exception
  }
```

File: deeplearning4j-scaleout/spark/dl4j-spark/src/main/spark-1/org/deeplearning4j/spark/util/BaseDoubleFlatMapFunctionAdaptee.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+import java.util.Iterator;
+
+/**
+ * DoubleFlatMapFunction adapter to hide incompatibilities between Spark 1.x and Spark 2.x
+ *
+ * This class should be used instead of direct referral to DoubleFlatMapFunction
+ *
+ */
+public class BaseDoubleFlatMapFunctionAdaptee<T> implements DoubleFlatMapFunction<T> {
```
Comment (huitseeker) [inline deeplearning4j-scaleout/spark/dl4j-spark/src/main/spark-1/org/deeplearning4j/spark/util/BaseDoubleFlatMapFunctionAdaptee.java]: It's deeply unsatisfying how un-generic this is, simply for a return type's member. Do you feel like factoring the `BaseUFlatMapFunctionAdapter<U, T> Implements UFlatMapFunction<U, T>` ?

File: deeplearning4j-scaleout/spark/pom.xml [hunk 3]
```diff
@@ -37,11 +37,33 @@
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
+        <spark.major.version>1</spark.major.version>
+        <spark.version>1.6.2</spark.version>
+        <hadoop.version>2.2.0</hadoop.version>
```
Comment (elyast) [inline deeplearning4j-scaleout/spark/pom.xml]: There was hadoop.version 1 before which I thought was a bit ancient and I bumped it to the version spark 1.6/1.5 have

File: deeplearning4j-scaleout/spark/pom.xml [hunk 4]
```diff
@@ -101,4 +117,35 @@
...
+        <profile>
+            <id>cdh5</id>
+            <repositories>
+                <repository>
+                    <id>org.apache.hadoop</id>
+                    <url>https://repository.cloudera.com/artifactory/cloudera-repos/</url>
+                </repository>
+            </repositories>
+            <properties>
+                <hadoop.version>2.0.0-cdh4.6.0</hadoop.version>
```
Comment (elyast) [inline deeplearning4j-scaleout/spark/pom.xml]: Haven't really change a thing here, it was like this before

File: deeplearning4j-scaleout/spark/dl4j-spark/src/main/spark-1/org/deeplearning4j/spark/util/BaseDoubleFlatMapFunctionAdaptee.java (hunk 2 above)
Comment (elyast) [inline deeplearning4j-scaleout/spark/dl4j-spark/src/main/spark-1/org/deeplearning4j/spark/util/BaseDoubleFlatMapFunctionAdaptee.java]: I thought about it, however I noticed that Spark interfaces are completely unrelated, e.g. DoubleFlatMapFunction doesn't extend FlatMapFunction, etc. I'm not sure if we gain anything there tough.

File: deeplearning4j-scaleout/dl4j-streaming/src/main/spark-2/org/deeplearning4j/streaming/pipeline/spark/StreamingContextUtils.java (hunk 1 above)
Comment (elyast) [inline deeplearning4j-scaleout/dl4j-streaming/src/main/spark-2/org/deeplearning4j/streaming/pipeline/spark/StreamingContextUtils.java]: I agree however Spark 2.x doesn't have this method:
jssc.awaitTermination(timeout), so I would anyway duplicated it with timeout, I'm not sure if its worth it

File: deeplearning4j-scaleout/spark/dl4j-spark/src/main/spark-1/org/deeplearning4j/spark/util/BaseDoubleFlatMapFunctionAdaptee.java (hunk 2 above)
Comment (elyast) [inline deeplearning4j-scaleout/spark/dl4j-spark/src/main/spark-1/org/deeplearning4j/spark/util/BaseDoubleFlatMapFunctionAdaptee.java]: Ok done it, since we are control of FlatMapFunctionAdapter we could do it like this.

**Comment by elyast** [comment]
@huitseeker let me know if you want some more changes

File: buildmultiplescalaversions.sh [hunk 5]
```diff
@@ -4,4 +4,7 @@ set -eu
 mvn "$@"
 ./change-scala-versions.sh 2.11
 mvn "$@"
+./change-spark-versions.sh 2
```
Comment (huitseeker) [inline buildmultiplescalaversions.sh]: So, same thing as what I said [there](https://github.com/deeplearning4j/DataVec/pull/171#discussion_r100231068) basically : the current state of the repo is a default of scala 2.11, and this "buildmultiple" script should reflect that.

File: deeplearning4j-scaleout/dl4j-streaming/src/main/spark-2/org/deeplearning4j/streaming/pipeline/spark/StreamingContextUtils.java (hunk 1 above)
Comment (huitseeker) [inline deeplearning4j-scaleout/dl4j-streaming/src/main/spark-2/org/deeplearning4j/streaming/pipeline/spark/StreamingContextUtils.java]: Yes, the argument of `awaitTermination` disappears in 2.1.x. But since you only use `awaitTermination` in the case where you have no applicable `timeout` being passed, that's not really an issue, is it ? You could:
```
try {
  (timeout < 0) ? jssc.awaitTermination() : jssc.awaitTerminationOrTimeOut(timeout);
} catch (Exception e) {
  if (e instanceof InterruptedException) {
    // handle exception
  }
  ```

no?

File: deeplearning4j-scaleout/spark/dl4j-spark/src/main/spark-1/org/deeplearning4j/spark/util/BaseDoubleFlatMapFunctionAdaptee.java (hunk 2 above)
Comment (huitseeker) [inline deeplearning4j-scaleout/spark/dl4j-spark/src/main/spark-1/org/deeplearning4j/spark/util/BaseDoubleFlatMapFunctionAdaptee.java]: \o/ 👏

File: deeplearning4j-scaleout/spark/pom.xml (hunk 3 above)
Comment (huitseeker) [inline deeplearning4j-scaleout/spark/pom.xml]: Fair enough

File: deeplearning4j-scaleout/spark/pom.xml (hunk 4 above)
Comment (huitseeker) [inline deeplearning4j-scaleout/spark/pom.xml]: Yeah, I know. And your PR shouldn't be dealing with this. 😫

File: buildmultiplescalaversions.sh [hunk 6]
```diff
@@ -2,6 +2,10 @@
 set -eu
 ./change-scala-versions.sh 2.10
 mvn "$@"
+./change-spark-versions.sh 2
```
Comment (elyast) [inline buildmultiplescalaversions.sh]: pushed

File: deeplearning4j-scaleout/dl4j-streaming/src/main/spark-2/org/deeplearning4j/streaming/pipeline/spark/StreamingContextUtils.java (hunk 1 above)
Comment (elyast) [inline deeplearning4j-scaleout/dl4j-streaming/src/main/spark-2/org/deeplearning4j/streaming/pipeline/spark/StreamingContextUtils.java]: I agree we can avoid duplication for case when no timeout is passed, however we do need to duplicate it when timeout is passed right? so what do we gain?

Comment (huitseeker) [inline deeplearning4j-scaleout/dl4j-streaming/src/main/spark-2/org/deeplearning4j/streaming/pipeline/spark/StreamingContextUtils.java]: Do we ?  `awaitTerminationOrTimeOut` has the same signature in Spark 1.6, & 2.x. AFAICT

https://spark.apache.org/docs/2.1.0/api/java/org/apache/spark/streaming/StreamingContext.html#awaitTerminationOrTimeout(long)
https://spark.apache.org/docs/1.6.2/api/java/org/apache/spark/streaming/StreamingContext.html#awaitTerminationOrTimeout(long)

Comment (elyast) [inline deeplearning4j-scaleout/dl4j-streaming/src/main/spark-2/org/deeplearning4j/streaming/pipeline/spark/StreamingContextUtils.java]: sorry my bad haven't checked it carefully :blush:

**Comment by huitseeker** [comment]
Thanks a lot ! \o/

**Comment by elyast** [comment]
:)
